### ISSUE TITLE
uunshare and uchroot: some nspawn-like proposal

### ISSUE THREAD

**Issue by solbjorn** [voidlinux/xbps#293]
Two questions I've got while migrating from systemd-based distro to Void:

1) Do we need an option to specify chroot hostname different from system one? I mean, something like `xbps-uunshare -h|--host buildoid [...]`. It can be useful in several cases, especially if user have 2-3-more chroot-rootfses, maybe with different architectures or/and for cross-compiling, and want to distinguish them that way. E.g. Linux Kbuild system writes compile-time hostname to _compile.h_ that becomes a part of runtime _/proc/version_ later.
2) Do we need an ability to bind-mount not only directories, but also single files? I'd rather bind _resolv.conf_ than manually copy and keep it in sync.

_(should mention that systemd-nspawn does not have options for that, it changes hostname and binds resolv.conf unconditionally)_

There's no problem for me to modify xbps sources and locally rebuild it via xbps-src, only want to know if there is any interest in getting this to upstream, so I could gladly open pull requests with described done.

**Comment by solbjorn** [comment]
UPD: one more quest.

3. Do we need an option to skip setgroup restriction setup for uunshare? Some utils like fakeroot or e.g. makedevs (from Buildroot project) use chown which becomes broken inside xbps-uunshared environment.

**Comment by Gottox** [comment]
1. It would be nice to set the hostname of the chroot.  My main usecase is reproducable builds
2. It's not exaclty needed but a nice addition.
3. I'd like to remove such restrictions too. My idea is to advertise uchroot/uunshare as a helper util for setting up news roots (`xbps-uchroot ... /mnt` instead of `mount --bind /dev /mnt/dev; mount --bind ...; chroot /mnt`

**Comment by Duncaen** [comment]
I'm not sure if we should add features that can't be added to both `xbps-uunshare` and `xbps-uchroot`.

And disabling `setgroups` is necessary for unprivileged namespaces to use `/proc/self/gid_map`.

**Comment by Gottox** [comment]
At least 1. is something that can be really useful for xbps-src.

**Comment by solbjorn** [comment]
@Duncaen, the first two opts would be added to both utils anyway. And the third is all about making chown work inside uunshared jails (there are no such problems in case of uchroot). So if there's any other ways to fix it, I'd rather use them than touch user namespaces mechanics.

**Comment by Duncaen** [comment]
I think only 2 is something that can be added.

How would you change the hostname in a `chroot(2)`?

Making `chown` work inside `xbps-uunshare` would mean you need setuid/root, it was added as an alternative for `xbps-uchroot` that does work without setuid/root.

I don't think we should make those too helpers alternatives for `systemd-nspawn`, they are two small helpers primarily for `xbps-src` and there are plenty of tools that allow you to create privileged and unprivileged namespaces with all the flexibility and options you might need. (bubblewrap, runc, containers, pflask...)

**Comment by ebfe** [comment]
There are a lot of more things to consider to make chown work in (unpriviledged) xbps-uunshare. If you are ok with a priviledged tool just use xbps-chroot. IMO if you want general purpose container tool use one of the already existing ones.

**Comment by solbjorn** [comment]
> How would you change the hostname in a `chroot(2)`?

Will `sethostname()` not work after clone syscall (as I see from uchroot source code, UTS NS is also cloned)?

Generally no, I don't want to overbloat our xbps helpers, of course. Since fakeroot is very special case (which hovewer is made for working without root privileges), yep, let my third question be not quite correct.

**Comment by Duncaen** [comment]
You are right, `CLONE_NEWUTS` is set in `xbps-uchroot` and would allow to change the hostname.
And because `xbps-uchroot` drops privileges it would be necessary to set the hostname early and can't be done in pid1 of the chroot. For `xbps-uunshare` it wouldn't be necessary but I think its a good idea to add the flag to both.

**Comment by solbjorn** [comment]
Great, will do both 1 & 2 in nearest week.
